ISPN-15060 Provide a proxable Cache API to get create queries

fax4ever opened the pull request:
https://issues.redhat.com/browse/ISPN-15060

~~Draft => Before to finish the implementation, I need to double-check with @tristantarrant if this was [the idea he described on the chat](https://infinispan.zulipchat.com/#narrow/stream/118645-infinispan/topic/Query/near/375957149).~~ done!

tristantarrant commented on core/src/main/java/org/infinispan/util/logging/Log.java:
```diff
@@ -2368,4 +2368,9 @@ CacheConfigurationException storeConfiguredHasBothReadAndWriteOnly(String storeC
    @Message(value = "Cache '%s' has number of owners %d but is missing too many members (%d/%d) to reinstall topology", id = 694)
    MissingMembersException missingTooManyMembers(String cacheName, int owners, int missing, int total);
+
+   @Message(value = "Embedded queries are not supported without adding any Infinispan query module to the classpath." +
```
"Query module not found. Add `infinispan-query` or `infinispan-query-core` to the classpath"

tristantarrant commented on commons/all/src/main/java/org/infinispan/commons/api/query/Query.java:
```diff
@@ -0,0 +1,132 @@
+    *  will be returned as the hit count in the {@link QueryResult}.
+    *
+    * @return {@link QueryResult} with the results.
+    * @since 11.0
```
Remove the @since from this method.
Add a @since 15.0 at the top of the interface

tristantarrant reviewed (changes requested):
Nice. I think we should deprecate the SearchFactory hinting at the "new" way. Please also work on the docs

fax4ever commented:
> Nice. I think we should deprecate the SearchFactory hinting at the "new" way. Please also work on the docs

Thanks for the review!
I applied the changes and also simplified the solution a little (I removed QueryProvider(s)).
About the deprecation I deprecated `QueryFactory#create(queryString)`, I don't know if we can deprecate the `Search#getQueryFactory`s since they contain other methods we're not replacing with the new API.

karesti commented on core/src/main/java/org/infinispan/cache/impl/SimpleCacheImpl.java:
```diff
@@ -923,6 +925,15 @@ public V remove(Object key) {
       return oldEntry != null ? oldEntry.getValue() : null;
    }
 
+   @Override
```
do we have some tests on this ?

fax4ever replied:
Great question. It seems that we don't have a test querying a simple cache. According to the documentation the simple cache cannot be indexed, but maybe it could be queryable...

fax4ever replied:
Actually it seems that a simple cache cannot be queryable with the current API. So I think we should do the same with the new API. I'm adding a change to present a proper message to the user.

fax4ever replied:
done! thank you @karesti

karesti reviewed (changes requested):
I really like this

karesti commented:
@fax4ever anything else needed ? docs only ?

fax4ever commented:
> @fax4ever anything else needed ? docs only ?

Yes. We need a feedback on the doc update only and I think that we can merge it.

karesti commented:
@fax4ever @domiborges do you have some time to review this or may we open a follow up jira for query docs ?

domiborges commented:
Hi! I'll check it in the afternoon

fax4ever commented:
> Hi! I'll check it in the afternoon

Thank you @domiborges

fax4ever commented:
wow 3 approvals :)

fax4ever commented:
Thank you @karesti
